### ISSUE TITLE
Using mprod with missing time points

### ISSUE THREAD

**Issue by sergioSEa** [UriaMorP/mprod_package#3]
Dear mprod development team,

First of all, I want to congratulate you for developing this package, it is really easy to use and it provides great biological insight to our projects.

I am writing this post mainly because I am not sure what is it happening when you include a dataset in which some of the samples are missing time points. The software is able to generate a solution (given that missing_flag is set to True), but I do not understand how is this affecting the final solution.

In the original manuscript it is mention that is recommended to impute the data upstream, however, when a whole time point (or more than one) consisting in hundreds of features are missing this becomes a daunting task.  

Could you provide me an intuition to understand how is the software dealing with this issue and how it may affect the resulting axis?

Thank you very much for your help!

**Comment by UriaMorP** [comment]
Hey @sergioSEa ,

I'm pleased to learn that you find the package easy to use and that it enables you to detect meaningful signals in your data.

**Before getting to answering what happens to your data, I would like to make it clear that:**

1. As you said 
> In the original manuscript it is mention that is recommended to impute the data upstream

This suggestion is due to the fact that the imputation of missing timepoints is a topic that stands on it's own, and one which we did not investigate as such in our paper.
2. The `missing_flag` argument of `table2tensor`  is an experimental feature meant for internal purposes and was unintentionally released with the software distribution. We will consider removing this feature or marking it as experimental in the next releases.
3. The current imputation method that is employed when `missing_flag == True` is described below. **We do not favor this method over any other imputation strategy, and do not have any educated opinion on the subject**

**To answer what is happening to your data when `missing_flag` is set to `True`:**

The `missing_flag` argument in `table2tensor` function simply masks the entries for which data was not obtained. Currently the only place where these values are being filled (with the mask value of 0) is in `MeanDeviationForm.transform` method. The operation is equivalent to zero imputation of entries of the data in its mean deviation form.

This means, that if you are following the workflow proposed e.g., in the docs, that is
```{python}
tensor, mode1_map, mode3_map = table2tensor(table_data, missing_samples = True)
tcam = TCAM()
transformed_data = tcam.fit_transform(tensor)    # invokes MeanDeviationForm.fit_transform()
```
then, after mean sample is computed (ignoring missing time-points) and subtracted from each horizontal slice, the missing values of the mean deviation form tensor are replaced with zeros. 

Concretely, suppose that your tensor $\cal{A}$ is missing the $i,j,k$'th entry (all other entries were obtained).  The mean sample of ${\cal A}$, is  $\bar{\mathbf{A}}=\frac{1}{m}\sum\_{i=1} {\cal A}\_{i,:,:}$ will ignore the missing entry (in this case $\bar{\mathbf{A}}\_{j,k} =  \frac{1}{m-1} \sum\_{\ell \neq i} {\cal{A}\_{\ell,j,k}}$ ).

The mean deviation form of $\cal{A}$ is defined by $\bar{\cal A}$ whose horizontal slices are $\bar{\cal A}\_{\ell,:,:} = \cal{A}\_{\ell,:,:} - \bar{\bf{A}}$ and this is the place where unmasking occurs: we have that $\bar{\cal A}\_{ijk} = 0$.

The rationale behind this method is that, in the "original space" of the input, missing values are replaced with mean observations.

Again, we do not claim this method is any better (or worse) than other strategies like forward/backward fill. Moreover, we strongly suggest that you explore all kinds of imputation methods and see which gets you the most logical and sensible results. 


> when a whole time point (or more than one) consisting in hundreds of features are missing this becomes a daunting task.

Consider using [pandas interpolation methods](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html), they are really easy to use :).


I am closing this issue now since it is outside of the package's intended scope. However, feel free to contact me via email - I will be more than happy to help in any way that I can (your can find my email address at the Elinav lab's website)
